util/codec/codec.go

```diff
@@ -734,13 +734,26 @@ func DecodeRange(b []byte, size int) ([]types.Datum, []byte, error) {
 		values = make([]types.Datum, 0, size)
 	)
 
+	i := 0
 	for len(b) > 1 {
 		var d types.Datum
-		b, d, err = DecodeOne(b)
+		if idxColumnType == nil {
+			b, d, err = DecodeOne(b)
+		} else {
+			if i >= len(idxColumnType) {
+				return values, b, errors.New("invalid length of index's columns")
+			}
+			if idxColumnType[i] == mysql.TypeDatetime {
```

Comment:
`mysql.TypeTimestamp` and `mysql.TypeDate` should be also decoded as `datetime`.
And please add some tests for these types.
@tangwz